File: src/converter/plugins/convertPrimitive.ts
import {ConverterPlugin, createSimplePlugin} from "../plugin.js";
import {Node} from "typescript";
import {Render} from "../render.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export function convertPrimitive<TNode extends Node = Node>(
    predicate: (node: Node) => node is TNode,
    render: Render<TNode>
): ConverterPlugin

export function convertPrimitive(
    predicate: (node: Node) => boolean,
    render: Render
): ConverterPlugin

export function convertPrimitive(
    predicate: (node: Node) => boolean,
    render: Render
): ConverterPlugin {
    return createSimplePlugin((node, context) => {
        if (!predicate(node)) return null

        const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
        checkCoverageService?.cover(node)

        return render(node)
    })
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertPrimitive.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.ConverterPlugin
import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node

fun convertPrimitive(
    predicate: (node: Node) -> Boolean,
    render: (render: Node) -> String?,
): ConverterPlugin<Node> {
    return createSimplePlugin plugin@{ node: Node, context, _ ->
        if (!predicate(node)) return@plugin null

        val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
        checkCoverageService?.cover(node)

        render(node)
    }
}
